r-reproducibility.Rmd

```diff
@@ -0,0 +1,1211 @@
+
+You load a package by writing:
+
+```r
+library(tidyverse)
+```
+
+Working with multiple R scripts and files, it quickly gets tedious to always
+write out each library function at the top of each script. A better way of
+managing this is to create a new file, keep all package loading code in that
+file, and sourcing that file in each R script. So, to create a new R file
+in the `R/` folder, we'll use this `use_r()` function from the usethis package:
+
+```r
+usethis::use_r("package-loading")
+```
+
+This creates a file called `package-loading.R` in the `R/` folder. In the
+`package-loading.R` file, add this code to it.
+
+```r
+library(tidyverse)
+```
+
+Then in other R scripts in the `R/`, include this code at the top the script:
+
+```r
+source(here::here("R/package-loading.R"))
+```
+
+There's a new thing `here`! The [here][r-here-pkg] package uses a function
+called `here()` that makes it easier to manage file paths. What is a file path
+and why is this necessary? A file path is the list of folders a file is found
+in. For instance, your resume may be found in
+`/Users/Documents/personal_things/resume.docx`. The problem with file paths in R
+is that when you run a script interactively (e.g. what we do in class and
+normally), the file path is located at the Project level (where the `.Rproj`
+file is found). You can see the file path by looking at the top of the
+"Console".
+
+But! When you `source()` an R script, it may likely run *in the
+folder it is saved in*, e.g. in the `R/` folder. So your file path
+`R/packages-loading.R` won't work because there isn't a folder called R in the
+`R/` folder. Often people use the function `setwd()`, but this is *never* a good
+idea since using it makes your script *runnable only on your computer*... which
+makes it no longer reproducible. We use the `here()` function to tell R to go
+to the project root (where the `.Rproj` file is found) and then use the file
+path from that point. This simple function can make your work more reproducible
+and easier for you to use later on.
+
+We also use the `here()` function when we import a dataset or save a dataset.
+So, let's load in the NYC Dog License dataset. First, save the csv in the
+`data/` folder. Then create a new R file:
+
+```r
+usethis::use_r("load-data")
+```
+
+And write these lines in the file:
+
+```{r load-data-example, eval=FALSE}
+source(here::here("R/package-loading.R"))
+dog_license <- read_csv(here::here("data/nyc-dog.....csv"))
+# TODO complete this file path.
+head(dog_license)
+```
+
+That is how we will load data in from now on.
+
+Here are a few other tips for keeping your code reusable:
+
+- Aggressively try keep your code as simple and concise as you can. The less code
+you have to write, debug, and maintain, the better. If you repeat some code,
+stop and think how to not repeat it. You will save time in the end by taking
+time to think and plan.
+- When encountering a difficult problem, think about how to solve it, then try
```

Comment:
Again, very high level advice tends not to be actionable. Would you recommend that people *not* think about how to solve it? If not, then recommending that they do may be redundant...